### PR TITLE
Firmware directory + code rearchitecture

### PR DESCRIPTION
**Pull request by jpan127**

### CODE REVIEW

File: firmware/main/L1_io/motor.h [hunk 1]
```diff
@@ -26,10 +26,16 @@ typedef enum
 {
     motor_dir_both_forward = 0,
     motor_dir_both_backward,
-    motor_dir_a_forward,    ///< A = left
+    motor_dir_a_forward,    ///< A = left @TODO : Fix this
```
Comment (jpan127) [inline firmware/main/L1_io/motor.h:29]: Captured in issue #16 .

File: firmware/main/L1_io/src/infrared.c [hunk 2]
```diff
@@ -9,75 +9,71 @@ static gpio_E infrared_map[infrared_max] = { 0 };
...
-    static const float slope = -41.4;
-    static const float y_intercept = 4963;
-
-    const float distance_cm = (voltage - y_intercept) / slope;
+    // Calibration parameters, characterized offline
+    static const float A02_slope = -41.4f;
+    static const float A02_y_intercept = 4963.0f;
 
-    return distance_cm;
+    /// @TODO : [#12] Additional characterization needed
```
Comment (jpan127) [inline firmware/main/L1_io/src/infrared.c:61]: Captured in #12 .

File: firmware/main/L1_io/src/motor.c [hunk 3]
```diff
@@ -60,80 +70,120 @@ bool motor_init(const motor_E motor, const motor_config_S * config)
 
 void motor_move(motor_E motor, motor_direction_E direction, float duty)
 {
+    // The right motor has a slight hardware-offset that causes the motors to move out of sync if using the same duty
+    static const float right_motor_negative_offset = 5.0f;
```
Comment (jpan127) [inline firmware/main/L1_io/src/motor.c]: Remove static, not efficient enough.

File: firmware/main/L2_app/packet.h [hunk 4]
```diff
@@ -36,12 +36,11 @@ typedef struct
     uint32_t rx_packets;                                ///< Number of received packets
     uint32_t tx_packets;                                ///< Number of transmitted packets
     uint32_t packets_dropped;                           ///< Number of packets dropped
-    uint32_t packet_counts[packet_type_last_invalid];   ///< Number of packets for each type
+    uint32_t packet_counts[packet_type_log_infrared];   ///< Number of packets for each type
```
Comment (jpan127) [inline firmware/main/L2_app/packet.h]: Revert.

File: firmware/main/L2_app/packet_structure.h [hunk 5]
```diff
@@ -22,7 +30,7 @@ typedef enum
     packet_type_log_wmark = 15,
     packet_type_log_infrared = 16,
 
-    packet_type_last_invalid,
+    packet_type_last_invalid = 20,
```
Comment (jpan127) [inline firmware/main/L2_app/packet_structure.h]: Revert.

File: firmware/main/L2_app/packet_structure.h [hunk 6]
```diff
@@ -43,8 +51,14 @@ typedef enum
...
     packet_opcode_stop = 14,
+    packet_opcode_deliver = 15,
+    packet_opcode_pivot_left = 16,
+    packet_opcode_pivot_right = 17,
+    packet_opcode_change_state = 18,
+    packet_opcode_execute_self_test = 19,
+    packet_opcode_set_suspend = 20,
 
-    packet_opcode_last_invalid,
+    packet_opcode_last_invalid = 21,
```
Comment (jpan127) [inline firmware/main/L2_app/packet_structure.h]: Revert.

File: firmware/main/L3_tasks/tasks/task_detection.c [hunk 7]
```diff
@@ -20,50 +57,164 @@ static bool infrared_self_test(void)
...
     }
 
     return success;
 }
 
+/**
+ *  Interprets the most current IR sensor readings
+ *  Acts appropriately, depending on which sensors triggered
+ */
+static void infrared_analyze_readings(void)
```
Comment (jpan127) [inline firmware/main/L3_tasks/tasks/task_detection.c:74]: Need to redo this function.

File: firmware/main/L3_tasks/tasks/task_detection.c [hunk 8]
```diff
@@ -20,50 +57,164 @@ static bool infrared_self_test(void)
...
+            //         sensor[infrared_top_right].trip_count = 0;
+            //     }
+            // }
+    }
+}
+
 void task_detection(task_param_T params)
 {
-    // Wait a second to make sure sensors are powered and have had some time to stabilize
+    // Wait a second to make sure sensors are powered and have had some time to stabilize @TODO: Remove because self test will have ample time
```
Comment (jpan127) [inline firmware/main/L3_tasks/tasks/task_detection.c:156]: Need to make sure.

File: firmware/main/L3_tasks/tasks/task_tx.c [hunk 9]
```diff
@@ -58,9 +61,9 @@ static bool open_socket_with_timeout(int *client_socket, uint32_t port, uint8_t
 static bool connect_socket_with_timeout(int *client_socket, uint32_t port, uint8_t task_id)
 {
     /// @TODO : Perfect place for leaky bucket
```
Comment (jpan127) [inline firmware/main/L3_tasks/tasks/task_tx.c:63]: Add leaky bucket module : #18.

File: state_machine.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
+from time import sleep
```
Comment (jpan127) [inline state_machine.py]: Delete this file.

File: firmware/main/L1_io/motor.h (hunk 1 above)
Comment (jpan127) [reply firmware/main/L1_io/motor.h:29]: Done.

File: firmware/main/L1_io/src/motor.c (hunk 3 above)
Comment (jpan127) [reply firmware/main/L1_io/src/motor.c]: Done.

File: firmware/main/L2_app/packet.h (hunk 4 above)
Comment (jpan127) [reply firmware/main/L2_app/packet.h]: Done.

File: firmware/main/L2_app/packet_structure.h (hunk 5 above)
Comment (jpan127) [reply firmware/main/L2_app/packet_structure.h]: Done.

File: firmware/main/L2_app/packet_structure.h (hunk 6 above)
Comment (jpan127) [reply firmware/main/L2_app/packet_structure.h]: Done.

File: firmware/main/L3_tasks/tasks/task_detection.c (hunk 8 above)
Comment (jpan127) [reply firmware/main/L3_tasks/tasks/task_detection.c:156]: If #13 is fixed, this will be sure.

File: firmware/main/L3_tasks/tasks/task_tx.c (hunk 9 above)
Comment (jpan127) [reply firmware/main/L3_tasks/tasks/task_tx.c:63]: Done.

File: state_machine.py (hunk 10 above)
Comment (jpan127) [reply state_machine.py]: Done.

File: firmware/main/L3_tasks/tasks/task_detection.c (hunk 7 above)
Comment (jpan127) [reply firmware/main/L3_tasks/tasks/task_detection.c:74]: Captured in #14 .

**Review by avirynd — APPROVED** [review]

**Pull request merged**